### PR TITLE
Add HTML + JS Support (POC)

### PR DESCRIPTION
**Pull request by GordonSmith**
Signed-off-by: Gordon Smith <gordonjsmith@gmail.com>

### CODE REVIEW

**Comment by GordonSmith** [comment]
@dehilsterlexis This is reference code only (and incomplete).  But it shows how to add a new language to the IDE and the scintilla editor (in this case its enabling HTML and JavaScript).  It is not for review nor ready for pulling.

File: CMakeLists.txt [hunk 1]
```diff
@@ -43,6 +43,12 @@ if (WIN32)
 	string(REGEX REPLACE "/Zm[0-9]+ *" "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
 	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Zm500" CACHE STRING "" FORCE)
 	ADD_DEFINITIONS( -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -DUNICODE -D_UNICODE )
+
+	STRING(REPLACE "/Zi" "/ZI" CMAKE_CXX_FLAGS_DEBUG ${CMAKE_CXX_FLAGS_DEBUG})
+	SET(CMAKE_SHARED_LINKER_FLAGS_DEBUG "${CMAKE_SHARED_LINKER_FLAGS_DEBUG} /SAFESEH:NO")
+	SET(CMAKE_MODULE_LINKER_FLAGS_DEBUG "${CMAKE_MODULE_LINKER_FLAGS_DEBUG} /SAFESEH:NO")
+	SET(CMAKE_EXE_LINKER_FLAGS_DEBUG "${CMAKE_EXE_LINKER_FLAGS_DEBUG} /SAFESEH:NO")
+
```
Comment (GordonSmith) [inline CMakeLists.txt]: This code enables edit and continue in the debugger with VS2015!

File: EclEditor/CMakeLists.txt [hunk 2]
```diff
@@ -4,6 +4,11 @@ set ( SRCS
 		lexecl.cxx
 		lexesdl.cxx
 
+		${SCINTILLA_INCLUDE_DIR}/include/SciLExer.h
+
+		${SCINTILLA_INCLUDE_DIR}/lexers/LexCPP.cxx
+		${SCINTILLA_INCLUDE_DIR}/lexers/LexHTML.cxx
+
```
Comment (GordonSmith) [inline EclEditor/CMakeLists.txt]: Add the new scintilla editors to the IDE project

File: EclEditor/Catalogue.cxx [hunk 3]
```diff
@@ -83,6 +83,9 @@ int Scintilla_LinkLexers() {
 //**\(\tLINK_LEXER(\*);\n\)
     LINK_LEXER(lmECL);
     LINK_LEXER(lmESDL);
+    LINK_LEXER(lmHTML);
+    LINK_LEXER(lmCPP);
+
```
Comment (GordonSmith) [inline EclEditor/Catalogue.cxx]: Register the new editors with scintilla.

File: comms/AttributeType.cpp [hunk 4]
```diff
@@ -304,6 +316,10 @@ bool IsValidExtension(const std::_tstring & ext)
         return true;
     else if (boost::algorithm::iends_with(ext, ATTRIBUTE_TYPE_CMP))
         return true;
+    else if (boost::algorithm::iends_with(ext, ATTRIBUTE_TYPE_HTML))
+        return true;
+    else if (boost::algorithm::iends_with(ext, ATTRIBUTE_TYPE_JS))
+        return true;
```
Comment (GordonSmith) [inline comms/AttributeType.cpp]: Add support for the new file extensions to the IDE

File: eclide/EclDlgBuilder.cpp [hunk 5]
```diff
@@ -14,6 +14,7 @@ CBuilderDlg::CBuilderDlg(IAttribute *attribute, IEclBuilderSlot * owner) : m_att
     {
         m_sigConn = m_attribute->on_refresh_connect(boost::ref(*this));
         m_attrMonitor = new CAttributeMonitor(m_attribute);
+        m_view.SetType(m_attribute->GetType());
```
Comment (GordonSmith) [inline eclide/EclDlgBuilder.cpp]: Allow switching of the editor "mode" at runtime, based off of the file type.

File: wlib/LangRef.cpp [hunk 6]
```diff
@@ -711,18 +727,147 @@ class CLangRef : public ILangRef, public CUnknown
...
+        else if (boost::algorithm::equals(type->GetRepositoryCode(), _T("js"))) {
+            g_langRef[type->GetRepositoryCode()] = new CJSLangRef();
+        }
+        else {
+            g_langRef[type->GetRepositoryCode()] = new CECLLangRef();
+        }
+        g_langRef[type->GetRepositoryCode()]->init();
+	}
+	return g_langRef[type->GetRepositoryCode()];
 };
```
Comment (GordonSmith) [inline wlib/LangRef.cpp]: Add new clases for each new known language (this is probably _not_ the way we will do it - I suspect we will instead have a new lang.xml file per language.

File: wlib/LangRef.h [hunk 7]
```diff
@@ -36,7 +38,7 @@ __interface ILangRef : public IUnknown
     void SetFontBold(int catID, bool bold);
 };
 
-WLIB_API ILangRef * CreateEclLangRef();
+WLIB_API ILangRef * CreateLangRef(IAttributeType * type);
```
Comment (GordonSmith) [inline wlib/LangRef.h]: Support more than just ECL

File: wlib/RepositoryTreeNode.cpp [hunk 8]
```diff
@@ -275,11 +275,12 @@ CAttributeNode::~CAttributeNode()
...
 {
 	m_DispText = m_displayQualified ? m_attribute->GetQualifiedLabel() : m_attribute->GetLabel();
-	std::_tstring lockedBy = m_attribute->GetLockedBy();
+    m_DispText += m_attribute->GetType()->GetFileExtension();
+    std::_tstring lockedBy = m_attribute->GetLockedBy();
 	if (lockedBy.length())
 	{
 		m_DispText += _T(" (");
-		m_DispText += lockedBy.c_str();
+		m_DispText += lockedBy;
```
Comment (GordonSmith) [inline wlib/RepositoryTreeNode.cpp]: Display the file extension in repository tree

File: wlib/SourceView.cpp [hunk 9]
```diff
@@ -345,16 +351,17 @@ void CSourceCtrl::Reformat()
...
     CString names;
-    for(int i = 1; i <= m_langRef->GetLangCatCount(); ++i)
+    for (int i = 1; i <= m_langRef->GetLangCatCount(); ++i)
         SetKeyWords(i - 1, m_langRef->GetLangNames(i, names));
 
     InitColors(m_langRef);
 
+    SetTabWidth(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_TAB_WIDTH));
+    SetUseTabs(!(bool)GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_TAB_USESPACES));
+
```
Comment (GordonSmith) [inline wlib/SourceView.cpp]: Switch to the language type at init time.

**Comment by GordonSmith** [comment]
Closing - @dehilsterlexis you can still reference this even when its closed

**Comment by dehilsterlexis** [comment]
@GordonSmith got it.

**Pull request closed without merging**